framework/core/instance.cpp

```diff
@@ -381,20 +381,54 @@ void Instance::query_gpus()
 
 PhysicalDevice &Instance::get_suitable_gpu()
 {
-	assert(!gpus.empty() && "No physical devices were found on the system.");
+	if (gpus.size() == 0)
+	{
+		throw std::runtime_error("No physical devices found.");
+	}
+
+	uint32_t count_device_type[VK_PHYSICAL_DEVICE_TYPE_CPU + 1];
+	memset(count_device_type, 0, sizeof(count_device_type));
+
+	for (uint32_t i = 0; i < gpus.size(); i++)
+	{
+		const auto deviceType = gpus[i]->get_properties().deviceType;
+		assert(deviceType <= VK_PHYSICAL_DEVICE_TYPE_CPU);
+		count_device_type[deviceType]++;
+	}
+
+	const VkPhysicalDeviceType device_type_preference[] = {
+	    VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU,
+	    VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU,
+	    VK_PHYSICAL_DEVICE_TYPE_VIRTUAL_GPU,
+	    VK_PHYSICAL_DEVICE_TYPE_CPU,
+	    VK_PHYSICAL_DEVICE_TYPE_OTHER};
 
-	// Find a discrete GPU
-	for (auto &gpu : gpus)
+	VkPhysicalDeviceType search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU;
```

Comment:
slight nit pick, `search_for_device_type` sounds like a `verb/action`, it's usually better to have `nouns/name` for variable names. (not really a rule though)
For example `search_for_device_type` should be something like `outcome_of_search_for_device`(not great name though but gets the point across) on the other hand `search_for_device_type` could be a good name for a function.

Similar concern with other variable names.

Reply:
@baryluk, any updates on this one?  I still see some questions about how to handle the ambiguity around "best" device - and an over-ride mechanism for those who want to choose what's best for them (separate PR?).  Also questions re: other comments above.  Thanks.